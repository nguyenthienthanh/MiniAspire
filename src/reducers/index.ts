import { combineReducers } from 'redux'
import loanReducers from './loan'
import loanRepaymentsReducer from './loanRepayments'
import loanRepaymentReducer from './loanRepayment'

const reducers = combineReducers({
  loan: loanReducers,
  loanRepayments: loanRepaymentsReducer,
  loanRepayment: loanRepaymentReducer,
})

export type ReduxReducers = Pick<
  ReturnType<typeof reducers>,
  'loan' | 'loanRepayments' | 'loanRepayment'
>

export default reducers
