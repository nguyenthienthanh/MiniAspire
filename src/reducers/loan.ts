import update from 'immutability-helper'
import {
  LoanReduxActionsType,
  FETCH_CURRENT_LOAN_REQUESTED,
  FETCH_CURRENT_LOAN_SUCCESS,
  FETCH_CURRENT_LOAN_FAILED,
  REPAY_LOAN_SUCCESS,
} from '../actions'
import { Loan } from '../types/loan'

export type LoanReduxState = {
  loading: boolean
  error?: Error
  data?: Loan
}

const initialState: LoanReduxState = {
  loading: false,
}

const loanReducer = (
  state: LoanReduxState = initialState,
  action: LoanReduxActionsType,
): LoanReduxState => {
  switch (action.type) {
    case REPAY_LOAN_SUCCESS:
      if (!state.data) {
        return state
      }

      return update(state, {
        data: {
          weeklyRepaid: {
            $set: true,
          },
        },
      })
    case FETCH_CURRENT_LOAN_REQUESTED:
      return update(state, {
        $set: {
          loading: true,
        },
      })
    case FETCH_CURRENT_LOAN_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loan,
        },
      })
    case FETCH_CURRENT_LOAN_FAILED:
      return update(state, {
        $set: {
          loading: false,
          error: action.error,
        },
      })

    default:
      return state
  }
}

export default loanReducer
