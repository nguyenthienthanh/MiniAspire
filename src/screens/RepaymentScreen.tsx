import React, { FC } from 'react'
import { View, Text, StyleSheet } from 'react-native'

export type RepaymentScreenProps = {}

const RepaymentScreen: FC<RepaymentScreenProps> = (props) => {
  return (
    <View>
      <Text />
    </View>
  )
}

export default RepaymentScreen

const styles = StyleSheet.create({})
