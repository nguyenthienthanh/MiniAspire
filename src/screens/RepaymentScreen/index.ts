export { default } from './RepaymentScreen'
export * from './RepaymentScreen'
