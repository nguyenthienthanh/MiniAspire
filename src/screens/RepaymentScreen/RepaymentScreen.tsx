import React, { FC, useCallback } from 'react'

import Container from '../../components/Container'
import { Card, Button } from 'react-native-elements'
import { useSelector, useDispatch } from 'react-redux'
import { ReduxReducers } from '../../reducers'
import { isEqual } from 'lodash'
import { repayLoan, fetchCurrentLoan } from '../../actions'

export type RepaymentScreenProps = {}

const RepaymentScreen: FC<RepaymentScreenProps> = (props) => {
  const loan = useSelector<ReduxReducers, ReduxReducers['loan']>(
    (state) => state.loan,
    isEqual,
  )

  const dispatch = useDispatch()

  React.useEffect(() => {
    if (!loan.data) {
      dispatch(fetchCurrentLoan())
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch])

  const loanRepayment = useSelector<
    ReduxReducers,
    ReduxReducers['loanRepayment']
  >((state) => state.loanRepayment, isEqual)

  const handleRepay = useCallback(() => {
    if (!loan.data?.id) {
      return
    }

    dispatch(repayLoan(loan.data.id))
  }, [dispatch, loan.data])

  return (
    <Container>
      <Card>
        <Button
          testID="repay"
          title="Repay"
          onPress={handleRepay}
          loading={loanRepayment.loading}
          disabled={loanRepayment.loading}
        />
      </Card>
    </Container>
  )
}

export default RepaymentScreen
