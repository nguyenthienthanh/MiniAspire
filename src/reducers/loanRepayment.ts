import update from 'immutability-helper'
import {
  LoanReduxActionsType,
  REPAY_LOAN_SUCCESS,
  REPAY_LOAN_REQUESTED,
  REPAY_LOAN_FAILED,
} from '../actions'
import { LoanRepayment } from '../types/loan'

export type LoanRepaymentReduxState = {
  loading: boolean
  error?: Error
  data?: LoanRepayment
}

const initialState: LoanRepaymentReduxState = {
  loading: false,
}

const loanRepaymentReducer = (
  state: LoanRepaymentReduxState = initialState,
  action: LoanReduxActionsType,
): LoanRepaymentReduxState => {
  switch (action.type) {
    case REPAY_LOAN_REQUESTED:
      return update(state, {
        loading: {
          $set: true,
        },
      })
    case REPAY_LOAN_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loanRepayment,
        },
      })
    case REPAY_LOAN_FAILED:
      return update(state, {
        loading: {
          $set: false,
        },
        error: {
          $set: action.error,
        },
      })

    default:
      return state
  }
}

export default loanRepaymentReducer
