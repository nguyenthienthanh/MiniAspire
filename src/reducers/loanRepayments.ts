import update from 'immutability-helper'
import {
  FETCH_LOAN_REPAYMENTS_FAILED,
  FETCH_LOAN_REPAYMENTS_REQUESTED,
  FETCH_LOAN_REPAYMENTS_SUCCESS,
  LoanReduxActionsType,
  REPAY_LOAN_SUCCESS,
} from '../actions'
import { LoanRepayment } from '../types/loan'

export type LoanRepaymentReduxState = {
  loading: boolean
  error?: Error
  data: LoanRepayment[]
}

const initialState: LoanRepaymentReduxState = {
  loading: false,
  data: [],
}

const loanRepaymentsReducer = (
  state: LoanRepaymentReduxState = initialState,
  action: LoanReduxActionsType,
): LoanRepaymentReduxState => {
  switch (action.type) {
    case REPAY_LOAN_SUCCESS:
      return update(state, {
        data: { $unshift: [action.loanRepayment] },
      })
    case FETCH_LOAN_REPAYMENTS_REQUESTED:
      return update(state, {
        loading: {
          $set: true,
        },
      })
    case FETCH_LOAN_REPAYMENTS_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loanRepayments,
        },
      })
    case FETCH_LOAN_REPAYMENTS_FAILED:
      return update(state, {
        $set: {
          loading: false,
          data: state.data,
          error: action.error,
        },
      })

    default:
      return state
  }
}

export default loanRepaymentsReducer
