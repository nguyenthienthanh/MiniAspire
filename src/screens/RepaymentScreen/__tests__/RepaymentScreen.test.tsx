/**
 * @format
 * @jest-environment jsdom
 */
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import 'react-native'
import * as Redux from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { ReduxReducers } from '../../../reducers'
import RepaymentScreen from '../RepaymentScreen'

Enzyme.configure({ adapter: new Adapter() })

const mockStore = configureStore()

describe('RepaymentScreen', () => {
  it('should render correctly and match the snapshot', () => {
    expect.assertions(2)

    const state: ReduxReducers['loanRepayment'] = {
      loading: true,
      data: { amount: 2000000, repaymentsAt: new Date() },
    }

    const wrapper = renderer.create(
      <Redux.Provider
        store={mockStore({
          loan: {
            loading: false,
            data: {
              id: 'loan-1',
              amount: 30000000,
              term: new Date(),
              weeklyAmount: 2000000,
              weeklyRepaid: false,
            },
          },
          loanRepayment: state,
        })}
      >
        <RepaymentScreen />
      </Redux.Provider>,
    )

    expect(wrapper.root.findByProps({ testID: 'repay' }).props.loading).toBe(
      true,
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
