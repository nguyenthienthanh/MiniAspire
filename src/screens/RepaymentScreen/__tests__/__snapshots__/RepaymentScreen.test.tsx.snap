// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepaymentScreen should render correctly and match the snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(242, 242, 242)",
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "paddingTop": null,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 15,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <View
          style={
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": true,
                "disabled": true,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="repay"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "hsl(208, 8%, 90%)",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <ActivityIndicator
                animating={true}
                color="white"
                hidesWhenStopped={true}
                size="small"
                style={
                  Object {
                    "marginVertical": 2,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
